### ISSUE TITLE
False positive for Array Init Violation when using map on Optional

### ISSUE THREAD

**Issue by LucasVanDongen** [realm/SwiftLint#2725]
### New Issue Checklist

- [V ] Updated SwiftLint to the latest version
- [V ] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

When I compile the following code, it will throw an Array Init Violation warning:
`let plate = self.unit.plate.map { $0 } ?? "\(self.unit.id)"`

Where `plate` is a `String?` and `id` is an `Int`

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.31.0
* Installation method used (Homebrew, CocoaPods, building from source, etc)? CocoaPods
* Paste your configuration file:

```yml
  - closure_spacing
  - collection_alignment
  - contains_over_first_not_nil
  - empty_count
  - empty_string
  - empty_xctest_method
  - explicit_init
  - extension_access_modifier
  - fallthrough
  - fatal_error_message
  - file_name
  - first_where
  - identical_operands
  - joined_default_parameter
  - let_var_whitespace
  - last_where
  - literal_expression_end_indentation
  - lower_acl_than_parent
  - nimble_operator
  - number_separator
  - object_literal
  - operator_usage_whitespace
  - overridden_super_call
  - override_in_extension
  - pattern_matching_keywords
  - private_action
  - private_outlet
  - prohibited_interface_builder
  - prohibited_super_call
  - quick_discouraged_call
  - quick_discouraged_focused_test
  - quick_discouraged_pending_test
  - redundant_nil_coalescing
  - redundant_type_annotation
  - single_test_class
  - sorted_first_last
  - sorted_imports
  - static_operator
  - unavailable_function
  - unneeded_parentheses_in_closure_argument
  - vertical_parameter_alignment_on_call
  - vertical_whitespace_closing_braces
  - vertical_whitespace_opening_braces
  - xct_specific_matcher
  - yoda_condition

identifier_name:
  excluded:
    - id
    - x
    - y
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
  If so, paste their relative paths and respective contents.
* Which Xcode version are you using (check `xcode-select -p`)? 10.1
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
        let plate: String? = "test"
        let unitId = 123
        let identifier = plate.map { $0 } ?? "\(unitId)"
```

**Comment by marcelofabri** [comment]
Why do you need the `map` in this case? Wouldn't `plate ?? "\(unitId)"` be equivalent?

**Comment by LucasVanDongen** [comment]
Absolutely true, but after editing some code and ending up with this I got this somewhat buggy behavior from the linter. I was just a bit surprised about the warning as it had nothing to do with what I was doing.

Perhaps it warrants a warning, but a different one

**Comment by marcelofabri** [comment]
Yeah, this rule is opt-in because we don't have enough information to avoid some false positives
